File: python/CustomSimpleCaptcha.py
__author__ = 'Tony Beltramelli - http://www.tonybeltramelli.com'

from random import randint
import json


class CustomSimpleCaptcha:
    _challenge_path = None
    _questions = None

    def __init__(self, challenge_path):
        self._challenge_path = challenge_path

    def get_challenge(self):
        questions = self._get_questions()

        index = randint(0, len(questions) - 1)

        challenge = questions[index]
        challenge = challenge[:challenge.index("=")]

        return json.dumps({'index': index, 'challenge': challenge})

    def check_answer(self, index, provided_answer):
        questions = self._get_questions()

        if (not isinstance(index, int)) or (index >= len(questions)):
            return self._get_status_message(2)

        expected_answer = questions[index]
        expected_answer = expected_answer[expected_answer.index("=") + 1:]

        is_matched = provided_answer.lower() == expected_answer.lower()

        return self._get_status_message(is_matched)

    def _get_questions(self):
        if self._questions is not None:
            return self._questions

        questions = open(self._challenge_path, "r")
        self._questions = questions.read().split("\n")
        self._questions = self._questions[1:len(self._questions)]

        return self._questions

    def _get_status_message(self, t):
        status = "error"

        if t == 0:
            status = "fail"
        elif t == 1:
            status = "success"

        return json.dumps({"status": status})

File: php/CustomSimpleCaptcha.php
<?php

/**
 * @author author Tony Beltramelli - http://www.tonybeltramelli.com
 */
 
class CustomSimpleCaptcha {
    
    private $_challengePath;
	private $_questions;
    
    public function __construct($challengePath) 
    {
    	$this->_challengePath = $challengePath;
    }
    
    public function getChallenge()
    {
       	$questions = $this->_getQuestions();
		
		$index = rand(0, count($questions) - 1);
		
		$challenge = $questions[$index];
		$challenge = substr($challenge, 0, strrpos($challenge, "="));
		
		$result = array("index" => $index, "challenge" => $challenge);
	
		return json_encode($result);
    }
	
	public function checkAnswer($index, $providedAnswer)
	{		
		$questions = $this->_getQuestions();
		
		if(!is_numeric($index) || $index >= count($questions)) return $this->_getStatusMessage(2);
		
		$expectedAnswer = $questions[$index];
		$expectedAnswer = substr($expectedAnswer, strrpos($expectedAnswer, "=") + 1, strlen($expectedAnswer));
		
		$isMatched = strcasecmp($providedAnswer, $expectedAnswer) == 0;
		
		return $this->_getStatusMessage($isMatched);
	}
	
	private function _getQuestions()
	{
		if(isset($this->_questions)) return $this->_questions;
		
		$questions = file_get_contents($this->_challengePath);
		$this->_questions = explode("\n", $questions);
		array_shift($this->_questions);
		
		return $this->_questions;
	}
	
	private function _getStatusMessage($type)
	{
		$status = "error";
		
		switch ($type)
		{
			case 0:
				$status = "fail";
				break;
			case 1:
				$status = "success";
				break;
			default:
				$status = "error";
		}
		
		$result = array("status" => $status);
		
		return json_encode($result);
	}
	
    public function __destruct() 
    {
    }
}

?>
